### PR TITLE
Convert from docbook to mdbook

### PR DESCRIPTION
**Pull request by noamraph**
Hi,

I find it significantly easier to read mdbook sites, such as the [Nix Reference Manual](https://nixos.org/manual/nix/stable/), than the current form of nix-pills. I also think that having the docs as markdown makes it much easier for others to contribute.

So I decided to try and convert nix-pills into mdbook. This PR is the result. The rendered HTML can be viewed at https://noamraph.github.io/nix-pills/. 

Most of the work was done by pandoc, and I added manual tweaks and fixes, as can be seen in the branch history.

What do you think?

Thanks,
Noam

TODO:
- [x] Make `nix-build release.nix -I nixpkgs=channel:nixos-unstable` build what's needed
- [x] Make old links continue to work

### CODE REVIEW

File: pills/02-install-on-your-running.xml
Comment (jtojnar) [inline pills/02-install-on-your-running.xml]: You are accidentally escaping the `<screen>` tags here.

File: pills/00-preface.md
Comment (jtojnar) [inline pills/00-preface.md:1]: Might be a good idea to follow https://github.com/NixOS/nixpkgs/issues/105243 for consistency with Nixpkgs manual.

Comment (noamraph) [reply pills/00-preface.md:1]: Thanks! I hope to try it in a few hours.

**Comment by kubukoz** [comment]
frankly, I'd approve this just for the removal of all the XML.

**Comment by noamraph** [comment]
@jtojnar I followed the suggestions you linked to. I converted all code sections into fenced sections, with the appropriate syntax highlighting. I think it looks very good.

See the result here: https://noamraph.github.io/nix-pills/

One thing which is missing is syntax highlighting for nix code. I understand this is done by using a version of highlight.js that supports nix, but this will wait for now.

Thanks,
Noam

File: scripts/perform_xml_include.py
Comment (DanielSidhion) [inline scripts/perform_xml_include.py:1]: Should this file get a comment explaining why it exists, or even be included in the PR? Seems like it was (very) useful for the migration, but could be confusing after the merge is done without at least a comment.

btw, amazing work!

File: pills/02-install-on-your-running.xml
Comment (noamraph) [reply pills/02-install-on-your-running.xml]: Fixed.

File: scripts/perform_xml_include.py
Comment (noamraph) [reply scripts/perform_xml_include.py:1]: Thanks! I agree, it's irrelevant to the updated repository. I kept it here https://github.com/noamraph/docbook-convert-stuff, for the tiny chance that I'll use it again, or perhaps someone else would find it useful.

**Comment by noamraph** [comment]
BTW, I now just took https://nixos.org/manual/nix/stable/highlight.js and put it in the gh-pages branch in my fork, and now nix code is highlighted - for example, see https://noamraph.github.io/nix-pills/08-generic-builders.html

**Comment by noamraph** [comment]
@infinisil thanks! I'm not sure what to do about the CI. I don't quite understand how the book is uploaded to https://nixos.org/guides/nix-pills/, and unfortunately I'm still a Nix newbie (that's why I converted the book, so it will be easier for me to learn...) so I don't quite understand how they work. I tried to write a `build.yml` file that would upload the book to my github-pages, but I tried to install mdbook using Nix and I didn't quite manage to make it work (See [here](https://github.com/noamraph/nix-pills/actions/runs/8559227411/job/23455544441)).

So, bottom line - I would love to get some help in making the CI build the book using mdbook!

**Comment by DanielSidhion** [comment]
@noamraph I'll help out here - will get to this tomorrow and make sure it outputs things as expected by the homepage. For reference, [here is what it expects](https://github.com/NixOS/nixos-homepage/blob/ce52d1bcdafe2394130cbb4786d3bbf2825b37ed/flake.nix#L169-L173).

In the meantime, have you used mdbook to generate an epub before? I know there's https://github.com/Michael-F-Bryan/mdbook-epub but I never used it. It would be a good idea to be able to still provide an epub output.

**Comment by infinisil** [comment]
I think it would be best to not bother with CI in this PR and instead just make sure that
```
nix-build release.nix -I nixpkgs=channel:nixos-unstable
```

works locally and produces the same structure. Keeping the same structure will probably be the trickiest bit, but it's very necessary because otherwise we break links.

**Comment by noamraph** [comment]
Thanks @DanielSidhion !

I managed to make the github action work (by manually adding `~/.nix-profile/bin` to the path), and then added epub generation using mdbook-epub. The result is at https://noamraph.github.io/nix-pills/nix-pills.epub , and I think it looks quite nice.

I also enabled Nix syntax highlighting - it turns out you just need to put `highlight.js` in a `theme` directory, which I took from the nix reference manual.

File: custom.css
Comment (jtojnar) [inline custom.css:18]: Missing newline.

Comment (jtojnar) [inline custom.css:18]: Would be nice to add no-copy prompts like in the Nix manual and Nixpkgs: https://github.com/NixOS/nixpkgs/pull/301439/files

File: .github/workflows/build.yml
Comment (infinisil) [inline .github/workflows/build.yml:53]: Unfortunately this won't work. As mentioned before, we really need `nix-build release.nix -I nixpkgs=channel:nixos-unstable` to work, so that the homepage can include it in its Nix build pipeline. We cannot use impure installation methods there, so this needs to be turned into a Nix build.

Comment (noamraph) [reply .github/workflows/build.yml:53]: Sure! I wanted to have something that works, I agree it shouldn't be committed like this. I hope to learn enough Nix to make by myself `nix-build` work, but if not, I guess it wouldn't be hard for others to help me with this.

I added checkboxes in the PR description for remaining tasks.

**Comment by noamraph** [comment]
@infinisil I added redirects (just lines in `book.toml`) from the old html names, such as `why-you-should-give-it-a-try.html`, to the new ones, such as `01-why-you-should-give-it-a-try.html`. Another alternative would have been to rename the markdown files to not have a preceding number, but I think this would have made editing significantly harder. (The XML files did have preceding numbers). So now this link works: https://noamraph.github.io/nix-pills/why-you-should-give-it-a-try

(Note: allowing links not ending with `.html` is a feature of github-pages, and it seems that also of the current nix-pills hosting. The generated files have a `.html` suffix)

How does this look to you?

File: custom.css
Comment (noamraph) [reply custom.css]: Nice! Done.

**Comment by jtojnar** [comment]
> Would it make sense to make the nix build script modify `book.toml` to have `follow-web-links = false` instead of `follow-web-links = true`, so when running outside of the sandbox external links would be checked?

Sounds like a good idea.

**Comment by DanielSidhion** [comment]
@jtojnar for what it's worth, I opened the epub on my reader and it "works". Not the greatest experience because some pages have almost no content in them, but at least the content seems to be there. I agree with what you said though. Unfortunately I don't have more time today to dedicate to this and try to update mdbook-epub on Nixpkgs, but if this is a blocker and is not resolved in the next days, I'll take a look at it early next week.

@noamraph yep! Sounds good to me. Can you make this change? It should be an extra command in the `buildPhase` of both outputs.

**Comment by jtojnar** [comment]
@DanielSidhion I will be able to continue in the evening if no one beats me to it.

**Comment by noamraph** [comment]
@DanielSidhion I'll try to make the change to modify `book.toml` by the build script!

@jtojnar I checked with the most recent version of mdbook-epub, and I get the same errors from epubcheck. As the epub is usable, I don't think this should block the change. Perhaps I'll have time to look into improving mdbook-epub...

By the way, a "diplomatic" question. The preface mentions this:

> an effort to port them to the current format was led by Graham Christensen (aka grahamc / gchristensen) and other contributors in 2017

Since the format is changed, this is not really correct. What fix would you suggest? (The most obvious fix I see would include my name, and it would be quite flattering to me, but I'm not sure if I deserve it. An alternative is "an effort to port them to the DocBook XML format was led by Graham Christensen and other contributors in 2017, and they were ported again to the current format in 2024". So, your call!)

**Comment by noamraph** [comment]
@jtojnar I managed to beat you :wink: I added the `sed` command, and now it will check external links outside of the sandbox.

**Comment by noamraph** [comment]
I removed the epub section from `book.toml`. The epub is still built by the standalone command.

It actually didn't work for me, due to some version warning. And I think it would be better if contributors could run `mdbook build` successfully without having to mess with `mdbook-epub`.

File: default.nix
Comment (jtojnar) [inline default.nix:18]: Use `substituteInPlace --replace-fail` to make it fail loudly if the `book.toml` is changed.

Comment (noamraph) [reply default.nix]: Done! I actually looked for a way to make sed fail if it doesn't find the pattern, but it wasn't trivial so I let it go.

File: default.nix
Comment (jtojnar) [inline default.nix:66]: If some viewer does display it, it does not mean others will be able to as well (they might have stricter parsers). Looks like there are two kinds of issues:

- In `OEBPS/content.opf` file, the `id` and `idref` attributes start with a number (since `mdbook-epub` derives them from a file name). Maybe we could work around it by using `pXX` prefix on files, or in the `content.opf`
- The interfile md links are not updated in text bodies.

The following workaround appears to work:

```patch
diff --git a/anchors.jq b/anchors.jq
new file mode 100644
index 0000000..7230977
--- /dev/null
+++ b/anchors.jq
@@ -0,0 +1,31 @@
+"\\[\\]\\{#(?<anchor>[^\\}]+?)\\}" as $empty_anchor_regex |
+"\\[(?<text>[^\\]]+?)\\]\\{#(?<anchor>[^\\}]+?)\\}" as $anchor_regex |
+
+
+def transform_anchors_html:
+    . | gsub($empty_anchor_regex; "<a name=\"" + .anchor + "\"></a>")
+      | gsub($anchor_regex; "<a href=\"#" + .anchor + "\" id=\"" + .anchor + "\">" + .text + "</a>");
+
+
+def transform_anchors_strip:
+    . | gsub($empty_anchor_regex; "")
+      | gsub($anchor_regex; .text);
+
+
+def map_contents_recursively(transformer):
+    . + {
+        Chapter: (.Chapter + {
+            content: .Chapter.content | transformer,
+            sub_items: .Chapter.sub_items | map(map_contents_recursively(transformer)),
+        }),
+    };
+
+
+def process_command:
+    .[0] as $context |
+    .[1] as $body |
+    $body + {
+        sections: $body.sections | map(map_contents_recursively(if $context.renderer == "html" then transform_anchors_html else transform_anchors_strip end)),
+    };
+
+process_command
diff --git a/book.toml b/book.toml
index 2e4902e..affaa23 100644
--- a/book.toml
+++ b/book.toml
@@ -33,6 +33,10 @@ git-repository-url = "https://github.com/NixOS/nix-pills"
 "fundamentals-of-stdenv.html" = "19-fundamentals-of-stdenv.html"
 "basic-dependencies-and-hooks.html" = "20-basic-dependencies-and-hooks.html"
 
+[preprocessor.anchors]
+renderers = ["html"]
+command = "jq --from-file anchors.jq"
+
 [output.linkcheck]
 follow-web-links = true
 
diff --git a/default.nix b/default.nix
index ae3ed31..1bee8f9 100644
--- a/default.nix
+++ b/default.nix
@@ -1,17 +1,15 @@
 { pkgs ? import <nixpkgs> {} }:
-let
-  buildInputs = with pkgs; [
-    mdbook
-    mdbook-epub
-    mdbook-linkcheck
-  ];
-in
+
 {
   html-split = pkgs.stdenvNoCC.mkDerivation {
     name = "nix-pills";
     src = ./.;
 
-    inherit buildInputs;
+    nativeBuildInputs = with pkgs; [
+      jq
+      mdbook
+      mdbook-linkcheck
+    ];
 
     buildPhase = ''
       # We can't check external links inside the sandbox, but it's good to check them outside the sandbox.
@@ -39,10 +37,29 @@ in
     name = "nix-pills-epub";
     src = ./.;
 
-    inherit buildInputs;
+    nativeBuildInputs = with pkgs; [
+      jq
+      mdbook-epub
+      unzip
+      zip
+    ];
 
     buildPhase = ''
       mdbook-epub --standalone .
+
+      mkdir nix-pills.epub-fix
+      ( cd nix-pills.epub-fix
+        unzip -q "../book/epub/Nix Pills.epub"
+        # Fix invalid ids.
+        sed -Ei 's/(id(ref)?=")([0-9])/\1p\3/g' OEBPS/content.opf
+        sed -Ei 's/(id="|href="#)([0-9])/\1fn\2/g' OEBPS/20-basic-dependencies-and-hooks.html
+        # Fix anchors, mdbook-epub does not run preprocessors.
+        sed -Ei 's~\[\]\{#([^\}]+)\}~<a id="\1" />~g' OEBPS/*.html
+        # Fix broken links in body.
+        sed -Ei 's/("[0-9a-z-]+\.)md(["#])/\1html\2/g' OEBPS/*.html
+        zip -q "../book/epub/Nix Pills.epub" **/*
+      )
+      # rm -r nix-pills.epub-fix
     '';
 
     installPhase = ''
@@ -65,18 +82,18 @@ in
     # TODO: It would be a great idea to run epubcheck and succeed, but since we moved to mdBook, the plugin that generates an epub creates an output that fails epubcheck in many ways that we can't fix ourselves.
     # In any case, the epub was manually inspected and has all the content displayed without big issues.
     # Because of this, we leave the code below as a comment so someone in the future can try enabling it again, or working on the mdBook epub plugin to fix the issues caught by epubcheck.
-    # installCheckInputs = with pkgs; [
-    #   epubcheck
-    # ];
+    installCheckInputs = with pkgs; [
+      epubcheck
+    ];
 
-    # doInstallCheck = true;
+    doInstallCheck = true;
 
-    # installCheckPhase = ''
-    #   runHook preInstallCheck
+    installCheckPhase = ''
+      runHook preInstallCheck
 
-    #   epubcheck "$manual"
+      epubcheck "$manual"
 
-    #   runHook postInstallCheck
-    # '';
+      runHook postInstallCheck
+    '';
   };
 }
diff --git a/pills/03-enter-environment.md b/pills/03-enter-environment.md
index 66b140c..dcb185e 100644
--- a/pills/03-enter-environment.md
+++ b/pills/03-enter-environment.md
@@ -1,4 +1,4 @@
-# Enter the Environment {#enter-environment}
+# Enter the Environment []{#enter-environment}
 
 Welcome to the third Nix pill. In the [second pill](02-install-on-your-running-system.md) we installed Nix on our running system. Now we can finally play with it a little, these things also apply to NixOS users.
 
@@ -16,7 +16,7 @@ $ source ~/.nix-profile/etc/profile.d/nix.sh
 
 To remind you, `~/.nix-profile/etc` points to the `nix-2.1.3` derivation. At this point, we are in our Nix user profile.
 
-## Install something
+## Install something []{#install-something}
 
 Finally something practical! Installation into the Nix environment is an interesting process. Let's install `hello`, a simple CLI tool which prints `Hello world` and is mainly used to test compilers and package installations.
 
diff --git a/pills/20-basic-dependencies-and-hooks.md b/pills/20-basic-dependencies-and-hooks.md
index 51dcac4..85a2769 100644
--- a/pills/20-basic-dependencies-and-hooks.md
+++ b/pills/20-basic-dependencies-and-hooks.md
@@ -273,4 +273,5 @@ and if one dependency has that setup hook then all of them will be so `echo`ed.
 ...I'm not sure! We could talk about the additional dependency types and hooks which cross compilation necessitates, building on our knowledge here to cover stdenv as it works today. We could talk about how nixpkgs is bootstrapped. Or we could talk about how `localSystem` and `crossSystem` are elaborated into the `buildPlatform`, `hostPlatform`, and `targetPlatform` each bootstrapping stage receives. Let us know which most interests you!
 
 [^1]: We can now be precise and consider what `addToEnv` does alone the minimal treatment of a dependency: i.e. a package that is _just_ a dependency would _only_ have `addToEnv` applied to it.
+
 [^2]: It was called [GCC Wrapper](https://github.com/NixOS/nixpkgs/tree/6675f0a52c0962042a1000c7f20e887d0d26ae25/pkgs/build-support/gcc-wrapper) in the version of nixpkgs suggested for following along in this pill; Darwin and Clang support hadn't yet motivated the rename.
```

Comment (noamraph) [reply default.nix:66]: @jtojnar do you want to apply this on the branch, so it will be recorded on your name, or do you prefer me to apply the patch?

**Review by jtojnar — COMMENTED** [review]
I found out that mdbook does actually support the [heading anchor override syntax](https://github.com/rust-lang/mdBook/pull/2013) but for some reason, it is not the case for mdbook-epub. So I dropped the jq script (since we do not use inline anchors) and handle the anchors in the epub derivation hack.

I have applied my previous suggestions as well as some other minor cleanups (e.g. fixing indentation of code blocks).

I have compared the epub outputs before and after and noticed the following differences:

- ❗ forced page breaks before headings
- ❗ pandoc strips `<package>` tags, maybe we should convert those to code like in Nixpkgs manual
- ❗ also `<replaceable>` inside `<filename>`
- it uses epub 2.0 instead of 3.0 
- headings not numbered
- TOC only show one level

I also scrolled the HTML outputs side by side and noticed the following additional changes:

- ❗ Admonitions are sometimes not properly spaced resulting in markdown inside not rendering – CommonMark requires a [blank line after opening HTML tag](https://spec.commonmark.org/0.31.2/#example-152). Example: Chapter 6
- In chapter 12, in the list at the bottom of the section 6 (The repository expression), options such as `-A` are wrapped in square brackets in DocBook version, no special marking in mdbook.

File: custom.css
Comment (noamraph) [reply custom.css:25]: I actually already solved it (by removing `.prompt_`), in a commit that doesn't exist in your force-push (e82be2c96e7b98d9521af006932df271413e2de0). Is there a reason for the force-push, rather than adding commits? I suggest that I restore the original branch, and then rebase your added commits on top of it. Is that OK with you?

File: pills/01-why-you-should-give-it-try.xml
Comment (noamraph) [reply pills/01-why-you-should-give-it-try.xml:136]: I found this: https://github.com/NixOS/nixpkgs/issues/175586#issuecomment-1144040425
I don't think there's real support for `replaceable`, but putting `«»` around the text seems like a reasonable replacement. Do you agree?

File: pills/18-nix-store-paths.xml
Comment (noamraph) [reply pills/18-nix-store-paths.xml:139]: Indeed! I will add backticks around it, once we decide what to do about the force push.

File: custom.css
Comment (jtojnar) [reply custom.css:25]: Updating highlight.js fixed it. I have force pushed because I have squashed some of the commits.

**Comment by noamraph** [comment]
I force-pushed, to restore the original commits. @jtojnar, I'm sorry for the extra work this causes you, but I couldn't really understand your force-push. Let's just keep normal pushes from now on, which may include "revert" pushes. I want to keep the history of this branch actually reflect the history.

The top commit of the branch was 09f40ff58d49141a0e3d94e8ee39168b0964ef29, which I pushed again.
On top of it, I added 3 commits which I just cherry picked from your commits.
There were conflicts with two commits, so I didn't cherry-pick them: a49ae81b9cfa7c69177126d5bd4ba40e503d1817, your latest commit, and d9218278cbefa8b8965a57d6b1801ea31d60e5ba. Can you re-apply the changes in those commits?

Thanks!
Noam

File: pills/18-nix-store-paths.xml
Comment (noamraph) [reply pills/18-nix-store-paths.xml:139]: Added in 6ddfb0a908142e49d518882e21ba40d8fb0239c0

File: pills/01-why-you-should-give-it-try.xml
Comment (noamraph) [reply pills/01-why-you-should-give-it-try.xml:136]: Done, commit a18780f9ecfe88d58e38a2442e91b62072a071c2

**Comment by jtojnar** [comment]
Thanks.

Re-applied the fixes using `git diff 09f40ff58d49141a0e3d94e8ee39168b0964ef29..c9240498febaa1d49d3e6ceffa11bd0735f41409 | git apply -3 -` and `git diff c9240498febaa1d49d3e6ceffa11bd0735f41409..55711cfcf3b50b80102afc042f8d3c57816bd9c7 | git apply -3 -`. Then `git reset` and staged individual hunks and created new commits.

I think this is in okay state now. Once you are happy, we should squash some of the commits so that the history does not contain dead ends (e.g. the non-Nix CI) and references to commits that are not in the repo (e.g. the cherry-picked commits. Though we should still take care to preserve the individual steps so it makes sense for people looking at `git blame` & co.

I suggest at least the following hunk reshuffling/squashes:

- “Remove "subtitle: Version" from 00-preface.md” (https://github.com/NixOS/nix-pills/commit/2fc7a33) became ”Remove front matter from 00-preface.md“ (https://github.com/NixOS/nix-pills/commit/eaa271c), also cherry picking the rest of removal from a later commit.
- Squash “Change build.yml to use mdbook” (https://github.com/NixOS/nix-pills/commit/a67830b), “Generate epub using mdbook-epub” (https://github.com/NixOS/nix-pills/commit/c6e620c), “Build the book with Nix” (https://github.com/NixOS/nix-pills/commit/d35f15a),
“Remove mdbook-epub from book.toml, to allow building without installing it” (https://github.com/NixOS/nix-pills/commit/4a5221e), “substitute sed with substituteInPlace, to fail early” (https://github.com/NixOS/nix-pills/commit/1ce7d90), “Fix a comment on how to run mdbook-epub” (https://github.com/NixOS/nix-pills/commit/09f40ff) and my amendment fixes mentioned above.
- Squash “Add redirects from original paths to new paths” (https://github.com/NixOS/nix-pills/commit/6311645) into “Fix inter-links which were broken due to filename changes” (https://github.com/NixOS/nix-pills/commit/b169693).
- Squash “Check external links when building outside the sandbox” (https://github.com/NixOS/nix-pills/commit/63bd775) the linkcheck enablement into “Support "note" sections, fix inter-links” (https://github.com/NixOS/nix-pills/commit/414e514), the Nix file change into the Nix build.

File: default.nix
Comment (jtojnar) [reply default.nix:58]: Work around in d833757854bda981e85995b7678f40e5548cef37

File: custom.css
Comment (jtojnar) [reply custom.css:25]: Updated again in e7a4eec962ed32551d34d5db19404051ed23b14b.

**Comment by noamraph** [comment]
@jtojnar ok, I squashed the commits as you suggested. The final result is exactly the same as was the previous, real, history.

(I feel that I wasted an hour of my life, but whatever 😅 )

Can you merge the PR?

**Comment by jtojnar** [comment]
Sorry, did not think it would be so bad.

Anyway, thank you, you did a great job.

**Comment by jtojnar** [comment]
Oh, looking at the nixos-homepage repo, this will still not build since we removed some arguments:

https://github.com/NixOS/nixos-homepage/blob/e691ed43624b2357acca1dcc86e740635e3dab20/flake.nix#L58-L62

@NixOS/documentation-team do you think it is meaningful to include build revision/date in the document, or should we remove it in nixos-homepage?

**Comment by DanielSidhion** [comment]
Huge thanks to both @noamraph and @jtojnar for your work to get a clean epub build :pray: 

I don't think it's meaningful to include the build revision and number (where we used the date as a proxy), since sometimes it sends the wrong signal about the "freshness" of the content both when the content is still relevant, but not updated for a while, and when the content is not relevant anymore, but another unrelated change makes it seem like it's fresh.

This PR had way too much churn, so I'll merge it and then make a PR on the homepage to remove those attributes. If there's a good argument to put the build revision and date back into the outputs, we can work that on a different PR.

**Review by DanielSidhion — APPROVED** [review]

**Comment by jtojnar** [comment]
Merged the PR manually to preserve the history.

**Comment by DanielSidhion** [comment]
Homepage PR is https://github.com/NixOS/nixos-homepage/pull/1385
Also created a PR to add credit for the current port: https://github.com/NixOS/nix-pills/pull/234

**Comment by noamraph** [comment]
> Sorry, did not think it would be so bad.
> 
> Anyway, thank you, you did a great job.

@jtojnar Thank you! It's OK, thank you for your hard work!

I realized that the history clean-up gives you less credit than you deserve. It happened because I cherry-picked to the working directory all the commits that I wanted to squash. Is there a way to preserve the credit next time?

Anyway, I appreciate your dedication!

**Comment by fricklerhandwerk** [comment]
Thanks everyone for making this land, great work!

**Comment by jtojnar** [comment]
> I realized that the history clean-up gives you less credit than you 
deserve. It happened because I cherry-picked to the working directory 
all the commits that I wanted to squash. Is there a way to preserve the 
credit next time?

You can include Co-authored-by trailer in the commit message. It is not 
as granular as individual commits but that is the tradeoff we pay for a 
clean history.

But I would not worry about it here. Most of the changes I made 
probably are not copyrightable and I am not doinf it for attribution 
anyway.

**Comment by noamraph** [comment]
@jtojnar thanks, I learned something!

**Pull request merged**
